ompi/mpi/c/finalized.c

```diff
@@ -48,17 +51,28 @@ int MPI_Finalized(int *flag)
                MPI_Finalize) or not */
 
             if (ompi_mpi_initialized && !ompi_mpi_finalized) {
+                opal_mutex_unlock(&ompi_mpi_bootstrap_mutex);
                 return OMPI_ERRHANDLER_INVOKE(MPI_COMM_WORLD, MPI_ERR_ARG,
                                               FUNC_NAME);
             } else {
+                opal_mutex_unlock(&ompi_mpi_bootstrap_mutex);
                 return OMPI_ERRHANDLER_INVOKE(null, MPI_ERR_ARG,
                                               FUNC_NAME);
             }
         }
     }
 
-    /* Pretty simple */
+    /* If some process has *started* FINALIZE but not yet completed
+       it, then be a little social and wait here until it is
+       completed. */
+    while (ompi_mpi_finalize_started && !ompi_mpi_finalized) {
+        opal_mutex_unlock(&ompi_mpi_bootstrap_mutex);
+        usleep(1);
+        opal_mutex_lock(&ompi_mpi_bootstrap_mutex);
+    }
 
```

Comment:
I specifically didn't hold the lock for the duration of ompi_mpi_init() / ompi_mpi_finalize().  Hence, I can't have the thread here just block waiting for the lock (e.g., which would naturally wait until the thread in ompi_mpi_init() completed the function).

Instead, I'm only using the lock to access the variables.  Hence, I have to loop around checking it, unlocking, delaying, and locking again.

Are you advocating that I should hold the lock through ompi_mpi_init() / ompi_mpi_finalize()?  I could probably do that (and then get rid of this loop, and the similar one in INITIALIZED)